drivers/flash/nxp_qspi_nor.c

```diff
@@ -0,0 +1,419 @@
+	flash_transfer.dataSize = 1;
+
+	do{
+		status = FLEXSPI_TransferBlocking(base_address, &flash_transfer);
+		if (status != kStatus_Success) {
+			return status;
+		}
+		// Status bit information for IS25LP064.
+		//
+		// The Write In Progress (WIP) bit is read-only, and can be used to
+		// detect the progress or completion of a program or erase operation. When the
+		// WIP bit is “0”, the device is ready for write status register, program or
+		// erase operation. When the WIP bit is “1”, the device is busy.
+		is_busy = ((read_value & 0x01UL) == 0x01UL) ? true : false;
+	} while (is_busy);
+
+	return status;
+}
+
+status_t flash_mcux_flexspi_qspi_write_enable(struct device *dev, off_t offset)
+{
+	status_t status;
+	flexspi_transfer_t flash_transfer;
+
+	struct flash_priv *priv = dev->driver_data;
+	FLEXSPI_Type *base_address = priv->pflash_block_base;
+
+	flash_transfer.deviceAddress = offset;
+	flash_transfer.port = kFLEXSPI_PortA1;
+	flash_transfer.cmdType = kFLEXSPI_Command;
+	flash_transfer.SeqNumber = 1;
+	flash_transfer.seqIndex = NOR_CMD_LUT_SEQ_IDX_WRITEENABLE;
+
+	status = FLEXSPI_TransferBlocking(base_address, &flash_transfer);
+
+	return status;
+}
+
+static int flash_mcux_flexspi_qspi_read(struct device *dev, off_t offset,
+					void *data, size_t len)
+{
+	status_t status;
+	flexspi_transfer_t flash_transfer;
+	unsigned int key;
+
+	key = irq_lock();
+
+	struct flash_priv *priv = dev->driver_data;
+	FLEXSPI_Type *base_address = priv->pflash_block_base;
+
+	flash_transfer.deviceAddress = offset;
+	flash_transfer.port = kFLEXSPI_PortA1;
+	flash_transfer.cmdType = kFLEXSPI_Read;
+	flash_transfer.SeqNumber = 1;
+	flash_transfer.seqIndex = NOR_CMD_LUT_SEQ_IDX_READ_FAST_QUAD;
+	flash_transfer.data = (uint32_t *)data;
+	flash_transfer.dataSize = len;
+
+	status = FLEXSPI_TransferBlocking(base_address, &flash_transfer);
+	if (status == kStatus_Success) {
+		status = flash_mcux_flexspi_qspi_wait_bus_busy(dev, offset);
+	}
+
+	irq_unlock(key);
+
+	return status;
+}
+
+#define FLASH_WRITE_SIZE 256
+
+static int flash_mcux_flexspi_qspi_write(struct device *dev, off_t offset,
+					 const void *data, size_t len)
+{
+	status_t status = kStatus_Fail;
+	flexspi_transfer_t flash_transfer;
+	unsigned int key;
+
+	static uint8_t ram_buffer[FLASH_WRITE_SIZE];
+
+	struct flash_priv *priv = dev->driver_data;
+	FLEXSPI_Type *base_address = priv->pflash_block_base;
+
+	if (k_sem_take(&priv->write_lock, K_NO_WAIT)) {
+		return -EACCES;
+	}
+
+	key = irq_lock();
+
+	int sectors = len / FLASH_WRITE_SIZE;
+
+	if (len % FLASH_WRITE_SIZE) {
+		sectors++;
+	}
+
+	for (int i = 0; i < sectors; i++) {
+		off_t offset_to_sector = offset + i * FLASH_WRITE_SIZE;
+		uint32_t *offset_to_data = (uint32_t *)((uint8_t *)data + i * FLASH_WRITE_SIZE);
+		uint32_t bytes_to_write = len - i * FLASH_WRITE_SIZE >= FLASH_WRITE_SIZE ? FLASH_WRITE_SIZE : len - i *
+					  FLASH_WRITE_SIZE;
+
+		memcpy(ram_buffer, offset_to_data, bytes_to_write);
```

Comment:
can we move source pointer here instead of memcpy?

Reply:
I guess we could, as the transfer operation is blocking anyway. I'll give it a try

Reply:
Do you have to worry about doing DMA from flash if the source buffer is not in SRAM?  If so the copy is unavoidable.